### ISSUE TITLE
About GATConv

### ISSUE THREAD

**Issue by BBDrive** [rusty1s/pytorch_geometric#2029]
## ❓ Questions & Help

Hello.  I have some questions about the source code of `GATConv`.
1. In the paper of `GAT`, when calculating self-attention, the embedding of the two nodes will be concatenated, I did not find this step in the code.
![image](https://user-images.githubusercontent.com/36299944/105144830-c832ca00-5b38-11eb-8dd0-6962cd8f4d6d.png)

**Comment by BBDrive** [comment]
I want to add edge features when calculating self-attention.

**Comment by ldv1** [comment]
in `forward`, you will find

```

if isinstance(x, Tensor):
            assert x.dim() == 2, 'Static graphs not supported in `GATConv`.'
            x_l = x_r = self.lin_l(x).view(-1, H, C)
            alpha_l = (x_l * self.att_l).sum(dim=-1)
            alpha_r = (x_r * self.att_r).sum(dim=-1)
        else:
            x_l, x_r = x[0], x[1]
            assert x[0].dim() == 2, 'Static graphs not supported in `GATConv`.'
            x_l = self.lin_l(x_l).view(-1, H, C)
            alpha_l = (x_l * self.att_l).sum(dim=-1)
            if x_r is not None:
                x_r = self.lin_r(x_r).view(-1, H, C)
                alpha_r = (x_r * self.att_r).sum(dim=-1)

```
Hence it is not implemented as you expect it to be; in the code, we compute for the left part (of a bipartite graph in general) the dot product of **a** with the embedding of **h_i**, then the same is done for the right part. The concatenation is thus a sum done in  `message`:

`alpha = alpha_j if alpha_i is None else alpha_j + alpha_i`

The idea is to make as few (redundant) operations in `message` as possible, and that's what the code achieved. At least, that's my interpretation.

So, if you want to add edge weights, I guess the alpha's must then be computed in `message`.

**Comment by rusty1s** [comment]
Yeah, that's correct. We avoid concatenation by just summing the individual parts in order to reduce memory usage. In case you want to integrate edge features, you may want to look at the `TransfomerConv`.

**Comment by BBDrive** [comment]
I get it. Thanks
